core/src/main/java/org/elasticsearch/bootstrap/BootstrapCheck.java

```diff
@@ -585,4 +586,32 @@ public String errorMessage() {
 
     }
 
+    static class DefaultClusterNameCheck implements Check {
+
+        private final ClusterName clusterName;
+
+        DefaultClusterNameCheck(final ClusterName clusterName) {
+            this.clusterName = clusterName;
+        }
+
+        @Override
+        public boolean check() {
+            return clusterName.equals(ClusterName.DEFAULT);
```

Comment:
> It wouldn't prevent misconfigured nodes from joining the cluster.

We can still check that the `cluster.name` key is present in the settings object, which means that it has been explicitly set, instead of letting the default override things. That would prevent unconfigured nodes from joining the cluster